examples/async-data/src/App.js

```diff
@@ -0,0 +1,186 @@
+import './App.css';
+import 'react-plotly.js-editor/lib/react-plotly.js-editor.css';
+import 'react-select/dist/react-select.css';
+import PlotlyEditor, {
+  EDITOR_ACTIONS,
+  Hub,
+  dereference,
+} from 'react-plotly.js-editor';
+import React, {Component} from 'react';
+import createPlotComponent from 'react-plotly.js/factory';
+import ee from 'event-emitter';
+import plotly from 'plotly.js/dist/plotly-basic';
+
+const Plot = createPlotComponent(plotly);
+
+class Backend {
+  constructor({dataSources, delay}) {
+    this.dataSources = dataSources;
+    this.delay = delay || 0;
+  }
+
+  fetch(request) {
+    const {resource, data} = request;
+    switch (resource) {
+      case 'GetChartingData':
+        this.processRequest('ChartingData', {
+          columnName: data.columnName,
+          data: this.dataSources[data.columnName],
+        });
+        break;
+      case 'GetChartingDataColumnNames':
+        this.processRequest(
+          'ChartingDataColumnNames',
+          Object.keys(this.dataSources)
+        );
+        break;
+      default:
+        throw new Error(`Unknown resource ${resource}`);
+    }
+  }
+
+  processRequest(resource, responseData) {
+    setTimeout(() => {
+      this.emit(resource, responseData);
+    }, this.delay);
+  }
+}
+
+ee(Backend.prototype);
+
+const backend = new Backend({
+  dataSources: {col1: [1, 2, 3], col2: [4, 3, 2], col3: [17, 13, 9]},
+  delay: 10,
+});
+
+class App extends Component {
+  constructor() {
+    super();
+
+    // A basic starting plotly.js figure object. Instead of assigning
+    const figure = {
+      data: [{type: 'scatter'}],
+      layout: {title: 'Room readings'},
+    };
+
+    // Store the figure, dataSource and dataSourceOptions in state.
+    this.state = {figure, dataSources: {}};
+
+    // Instantiate a Hub object. The hub is a convenience module that updates
+    // the applies Editor updates to the figure object. After an update it
+    // will call the onUpdate function with the editor and plot revision numbers.
+    // We set these in our state and pass them to react-plotly.js-editor and
+    // react-plotly.js plot component respectively. This is necessary because
+    // the plotly.js library will mutate the figure object with user interactions.
+    // The hub listens for events from plotly.js and alerts us to the mutation here.
+    // The Editor follows the same mutation pattern (for good or ill) and the Hub
+    // will pick up editor results in the same way.
+    this.hub = new Hub({
+      debug: true,
+      onUpdate: ({editorRevision, plotRevision}) =>
+        this.setState(prevState => ({
+          ...prevState,
+          editorRevision,
+          plotRevision,
```

Comment:
why does a user have to manage these things? Because we mutate data? I don't think that is a good enough excuse.